### DISCUSSION TITLE
How to defining Valibot schemas that agree with a TypeScript type?

### DISCUSSION THREAD

**Discussion by hungify** [open-circle/valibot#377] [Q&A] [answered]
Example from Zod

```
import { toZod } from 'tozod';

type Player = {
  name: string;
  age?: number | undefined;
  active: boolean | null;
};

export const Player: toZod<Player> = z.object({
  name: z.string(),
  age: z.number().optional(),
  active: z.boolean().nullable(),
});
```

How can I achieve that in Valibot?

**Comment by fabian-hiller** [comment, marked as answer]
You can use `GenericSchema`:

```ts
import * as v from 'valibot';

type Player = {
  name: string;
  age?: number | undefined;
  active: boolean | null;
};

export const Player = v.object({
  name: v.string(),
  age: v.optional(v.number()),
  active: v.nullable(v.boolean()),
}) satisfies v.GenericSchema<Player>;
```

**Reply by hungify** [reply]
Thanks a lot <3

**Reply by hungify2022** [reply]
@fabian-hiller 
This way has a limitation.
Is there any way to support an auto-complete object key?

**Reply by fabian-hiller** [reply]
Can you provide me with a code example?

**Reply by hungify2022** [reply]
@fabian-hiller 
With an example above.

https://github.com/fabian-hiller/valibot/assets/126773143/3abfe3e4-8b15-41ee-8be2-b48bd37b6169

**Reply by fabian-hiller** [reply]
Thank you. Now I get it. No, that does not work. This is the only way it would work, but it's way too complicated:

```ts
import * as v from 'valibot';

export const Player = v.object({
  name: v.string(),
  age: v.optional(v.number()),
  active: v.nullable(v.boolean()),
}) satisfies v.ObjectSchema<{
  name: v.StringSchema<string>;
  age: v.OptionalSchema<v.NumberSchema<number>, undefined>;
  active: v.NullableSchema<v.BooleanSchema<boolean>, undefined>;
}>;
```

**Reply by fabian-hiller** [reply]
Can you explain your use case to me? Why don't you use the schema as the source of true for validation and type?

**Reply by hungify2022** [reply]
Many thanks <3

I'm trying an integration type-safe API client with [openapi-ts](https://openapi-ts.pages.dev/introduction).
I want to ensure requests and responses are correct (another layer).

Your solution helped me create the correct schema easily.

**Reply by hungify2022** [reply]
I found another way to create a schema based on types. 
It matches with `v.BaseSchema<Player>` without `v.ObjectSchema<{...}>`

https://sinclairzx81.github.io/typebox-workbench/

**Reply by fabian-hiller** [reply]
Yes, that's a cool website!

**Reply by LexSwed** [reply]
Hey @fabian-hiller 👋 
I have another use case where my types are generated from protobufs ([example](https://github.com/improbable-eng/ts-protoc-gen/blob/master/examples/generated-grpc-web/proto/orphan_pb_service.d.ts)) and since every field in the proto definition is nullish, there needs to be a validation of the gRPC response for the clients to expect specific data structure used for rendering.
With the current `satisifies BaseSchema` solution, when a field is removed from the API, but it still exists in validation, TypeScript won't complain, due to `satisifes` being permissive to additional properties.
Maybe something like generic on `ObjectEntries` can help?

```ts
// autogenerated from proto
type Money = {
  currency: Currency;
  amount: number;
}

const moneySchema = object<Money>({
  currency: picklist(currencies),
  amount: number(),
  // here TypeScript should complain because extra no longer exists on Money type
  extra: number(),
})
```
I don't even want to validate that the output types are matching (e.g. `amount: number` not matching `amount: string`), but `extra` keys should fail in TS. Any ideas would be appreciated 🙏

**Reply by fabian-hiller** [reply]
Right now I do not have time to look into this, but I am aware that this is an important feature for many users. I will try to find a solution after v1 is released.

**Reply by LexSwed** [reply]
Thanks and no worries!

Just sharing a solution I landed on (with JSDoc explanation):
```ts
/**
 * Validates the output and keys of generated types for @type {object} to highlight
 * specific fields or keys that aren't matching the types.
 * Partial generics are not supported @see {@link https://github.com/microsoft/TypeScript/issues/26242},
 * so to provide a type, but also infer the @type {ObjectEntries}, using currying for now instead.
 */
const objectBySchema =
  <const TExpected extends Record<string, unknown>>() =>
  <const TEntries extends ObjectEntries>(
    // show error when there are no keys in common between defined schema and generated types
    schema: keyof TEntries & keyof TExpected extends never
      ? never
      : {
          // allow only the keys that are in the generated types
          [K in keyof TExpected & keyof TEntries]: InferOutput<
            TEntries[K]
          > extends TExpected[K]
            ? TEntries[K]
            : GenericSchema<TExpected[K]>;
        },
  ) =>
    object(schema);
```

**Reply by AgarwalPragy** [reply]
> You can use `BaseSchema`:
> 
> ```ts
> import * as v from 'valibot';
> 
> type Player = {
>   name: string;
>   age?: number | undefined;
>   active: boolean | null;
> };
> 
> export const Player = v.object({
>   name: v.string(),
>   age: v.optional(v.number()),
>   active: v.nullable(v.boolean()),
> }) satisfies v.GenericSchema<Player>;
> ```

@fabian-hiller That allows the schema to have more fields that what the `Player` type allows, even if the schema is a `v.strictObject` schema.

**Reply by AgarwalPragy** [reply]
@LexSwed here's a simplified version which doesn't require nested functions

```ts
import * as v from 'valibot'

/**
 * Utility to create a strict object valibot schema based on the expected type.
 *
 * @example
 * type User = { id: string; name: string; }
 * const UserSchema = strictObject<User>({ id: v.string(), name: v.string() })  // ✅
 * const MissingSchema = strictObject<User>({ id: v.string() })  // ❌ Missing 'name' property
 * const ExtraSchema = strictObject<User>({ id: v.string(), name: v.string(), extra: v.string() })  // ❌ Extra 'extra' property
 */
export const strictObject = <T extends Record<string, unknown>>(schema: { [K in keyof T]: v.GenericSchema<T[K]> }) =>
    v.strictObject(schema)
```
